etcdserver/etcdhttp/client_security.go

```diff
@@ -0,0 +1,267 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdhttp
+
+import (
+	"encoding/json"
+	"log"
+	"net/http"
+	"path"
+	"strings"
+
+	"github.com/coreos/etcd/etcdserver"
+	"github.com/coreos/etcd/etcdserver/etcdhttp/httptypes"
+	"github.com/coreos/etcd/etcdserver/security"
+)
+
+type securityHandler struct {
+	store       *security.Store
+	clusterInfo etcdserver.ClusterInfo
+}
+
+func handleSecurity(mux *http.ServeMux, server *etcdserver.EtcdServer) {
+	sec := securityHandler{
+		store:       security.NewStore(server, defaultServerTimeout),
+		clusterInfo: server.Cluster,
+	}
+	mux.HandleFunc(securityPrefix+"/roles", sec.baseRoles)
+	mux.HandleFunc(securityPrefix+"/roles/", sec.handleRoles)
+	mux.HandleFunc(securityPrefix+"/users", sec.baseUsers)
+	mux.HandleFunc(securityPrefix+"/users/", sec.handleUsers)
+}
+
+func (sec *securityHandler) handleRoles(w http.ResponseWriter, r *http.Request) {
+	subpath := path.Clean(r.URL.Path[len(securityPrefix):])
+	// Split "/roles/rolename/command".
+	// First item is an empty string, second is "roles"
+	pieces := strings.Split(subpath, "/")[2:]
+	switch len(pieces) {
+	case 1:
+		sec.forRole(w, r, pieces[0])
+	case 2:
+		sec.actionRole(w, r, pieces[0], pieces[1])
+	}
+}
+
+func (sec *securityHandler) baseRoles(w http.ResponseWriter, r *http.Request) {
+	if !allowMethod(w, r.Method, "GET") {
+		return
+	}
+	w.Header().Set("X-Etcd-Cluster-ID", sec.clusterInfo.ID().String())
+	w.Header().Set("Content-Type", "application/json")
+	var rolestruct struct {
+		Roles []string `json:"roles"`
+	}
+	roles, err := sec.store.AllRoles()
+	if err != nil {
+		writeError(w, err)
+		return
+	}
+
+	rolestruct.Roles = roles
+	err = json.NewEncoder(w).Encode(rolestruct)
+	if err != nil {
+		log.Println("etcdhttp: Error encoding on", r.URL)
+	}
+}
+
+func (sec *securityHandler) forRole(w http.ResponseWriter, r *http.Request, role string) {
+	if !allowMethod(w, r.Method, "GET", "PUT", "DELETE") {
+		return
+	}
+	w.Header().Set("X-Etcd-Cluster-ID", sec.clusterInfo.ID().String())
+
+	switch r.Method {
+	case "GET":
+		w.Header().Set("Content-Type", "application/json")
+		data, err := sec.store.GetRole(role)
+		if err != nil {
+			writeError(w, err)
+			return
+		}
+		err = json.NewEncoder(w).Encode(data)
+		if err != nil {
+			log.Println("etcdhttp: Error encoding on", r.URL)
+			return
+		}
+	case "PUT":
+		var in security.Role
+		err := json.NewDecoder(r.Body).Decode(&in)
+		if err != nil {
+			writeError(w, err)
+			return
+		}
+		if in.Role != role {
+			httptypes.NewHTTPError(400, "Role JSON name does not match the name in the URL")
+		}
+		err = sec.store.CreateRole(in)
+		if err != nil {
+			writeError(w, err)
+			return
+		}
+	case "DELETE":
+		err := sec.store.DeleteRole(role)
+		if err != nil {
+			writeError(w, err)
```

Comment:
if it is, it returns an etcd Error, with the appropriate semantics, which writeError already handles